[LIGO-577] Restructure

sashasashasasha151 opened the pull request:
[//]: # (This is a template of a pull request template.)
[//]: # (You should modify it considering specifics of a particular repository and put it there.)
[//]: # (Comments like this are meta-comments, they shouldn't be present in the final template.)
[//]: # (Comments starting with '<!---' are intended to stay in the final template.)

[//]: # (Keep in mind that it's only a template which contains items relevant to almost all conceivable repository.)
[//]: # (There can be other important items relevant to your repository that you can add here.)

## Description

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

[//]: # (Here you can add a link to the corresponding issue tracker, e. g. https://issues.serokell.io/issue/AD-)
[//]: # (For GitHub/GitLab issues it is better to use just hash symbol or exclamation mark as it is more resistant to repo movements)
[//]: # (In this case please also prefix the link with "Resolves" keyword)
[//]: # (See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
[//]: # (See https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords)
## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

[//]: # (Mostly for public repositories)
[//]: # (Recording changes is optional, depends on repository, useful for some libs)
- Public contracts
  - [ ] Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - [ ] I added an entry to the [changelog](../tree/master/CHANGES.md) if my changes are visible to the users
        and
  - [ ] provided a migration guide for breaking changes if possible

#### Stylistic guide (mandatory)

[//]: # (Update link to style guide if necesary or remove if it's not present)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

sashasashasasha151 committed:
[LIGO-577] Restructure

Problem: current structure of the Ligo IDE is more
like monorepo with a lot of packages. It was a good
solution for EthereumStudio as for the modular IDE.
But for our solution we do not need it at all as it
cause different problems on development stage.

Solution: remove such monorepo structure and put all
app code to the single package.

sashasashasasha151 committed:
[LIGO-577] Style and lint configs

Problem: currently our code doesn't match any style
and lint configs.

Solution: add proper eslint config and run `fix`
command on it

pgujjula reviewed (changes requested):
Can you update the build instructions in `ligo-webide-frontend/README.md`? I tried building with `yarn install` and then `yarn build:react` inside `ligo-webide-frontend/ligo-ide`, but the build failed.

Also, can you add a problem/solution-style commit message and description to the MR?

sashasashasasha151 committed:
[LIGO-577] Fix all style and lint issues

Problem: new config cause a lot of issues in our code
which we need to fix manually.

Solution: as there are a lot of problems for js files
we simply disabled all meaningful rules. For ts files
some problems were fixed and other one disabled

sashasashasasha151 committed:
[LIGO-577] Update readme

sashasashasasha151 committed:
[LIGO-577] Update typescript

pgujjula commented on ligo-webide-frontend/ligo-ide/.eslintrc.js:
```diff
@@ -0,0 +1,192 @@
+module.exports = {
```
Did you write this file, or import it from somewhere?

pgujjula replied:
If imported from a another licensed project we should put a note at the top about where it's from

sashasashasasha151 replied:
I used `edna` file and modified it (a lot), but, hm, it is like common thing, I mean it is just config, do we really need to do it?

gromakovsky replied:
AFAIK some files are not "copyrightable", maybe it's one of such files. But anyway, we have a policy about tracking licensing information on per-file basis, so ideally it should be done anyway and we should have a header here. This way licensing information is supposed to be not lost when you copy files. So if you copied it from https://github.com/serokell/edna/blob/master/frontend/.eslintrc.js, it should have the following header:
```
// SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io>
//
// SPDX-License-Identifier: AGPL-3.0-or-later
```

AFAIU we use `CC0-1.0` for files that are likely not "copyrightable". If you think it's such a file, I think we can write
```
// SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io>
//
// SPDX-License-Identifier: CC0-1.0
```

It's probably ok to change the license of files "owned" by us on the fly.

AFAIU neither `AGPL-3.0-or-later` nor `CC0-1.0` should cause any troubles. AGPL imposes some restrictions, but this particular file is not used in the build process, AFAIU it's used only for linting, so whatever gets deployed can be produced with `.eslintrc.js`. And having `SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io>` seems fine, because originally this file was written by us in the codebase owned by us.

sashasashasasha151 replied:
>originally this file was written by us in the codebase owned by us

Where is that edge when we should add license information? For example I also copied `.prettierrc` but it has only 5 lines of meaningful code and even more I think a lot of `.prettierrc` configs in the world has the same 5 lines of code.

>And having SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io> seems fine

I should add only this line or also `SPDX-License-Identifier: CC0-1.0`?

gromakovsky replied:
The idea is to add it to all files, so that `reuse lint` passes. But yes, there should be some "edge" between "copyrightable" and "non-copyrightable". AFAIU, the idea is that for "non-copyrightable" files you can always pretend that you wrote them from scratch, because there is no inellectual property in them. So you can write any license for them. And we use `CC0-1.0` for such files because AFAIU it means "do whatever you want".

I don't know whether this `.eslintrc.js` should be considered copyrightable to be honest. But in this case it probably doesn't really matter, because even if it's copyrightable, this copyright belongs to us, so we can do whatever we want with this file.

I think for non-copyrightable files it's not very useful to provide copyright information, but for example specifying `CC0-1.0` or another unlicense-like license you make it explicitly that you don't care about licensing for that file. Another reason to do it is to please automated tooling (`reuse lint`).

> I should add only this line or also `SPDX-License-Identifier: CC0-1.0`?

The whole header. AFAIU it's the format expected by `reuse lint`. And I think we want `reuse lint` to pass eventually.

pgujjula commented:
Copied from [Slack](https://serokell.slack.com/archives/C03CSCS30MR/p1659090943540179?thread_ts=1658365327.848129&cid=C03CSCS30MR) for sake of convenience and documentation:

I went through it and tested the refactored code, everything seems to be working well. One thing that might be a problem:
> All build/lint/package/license info from these packages was removed.

I notice that all the LICENSE files are just the text of the GPL, so I think it's fine to remove them. But I don't think we're allowed to remove the names of all the authors from the package.json files. I think we should document somewhere who the original package authors are. Especially because it seems that many of these are external packages that are not from Obsidian Labs.

EDIT: Resolved, see Slack for discussion

pgujjula reviewed (approved):
LGTM
